File: saco/2019/dna/dna.cpp
#pragma GCC optimize("O3,unroll-loops")
#pragma GCC target("avx,avx2,fma,bmi,bmi2")
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

map<char, int> mp = {{'A', 0}, {'C', 1}, {'T', 2}, {'G', 3}};

struct TrieNode
{
    bool is_end;
    TrieNode *children[4];
    TrieNode()
    {
        is_end = false;
        for (int i = 0; i < 4; i++)
            children[i] = nullptr;
    }
};

void insert(TrieNode *root, const string &s)
{
    TrieNode *p = root;
    for (char c : s)
    {
        int idx = mp[c];
        if (p->children[idx] == nullptr)
            p->children[idx] = new TrieNode();
        p = p->children[idx];
    }
    p->is_end = true;
}

int n, m;
string s;
int dp[5001];

int main()
{
    cin >> n >> m >> s;

    TrieNode *root = new TrieNode();
    while (m--)
    {
        int k;
        string t;
        cin >> k >> t;
        reverse(t.begin(), t.end());
        insert(root, t);
    }

    dp[0] = 0;
    for (int i = 1; i <= n; i++)
    {
        dp[i] = INT_MAX;

        TrieNode *p = root;
        for (int j = i - 1; j >= 0; j--)
        {
            p = p->children[mp[s[j]]];
            if (p == nullptr) break;

            if (p->is_end && dp[j] != INT_MAX)
                dp[i] = min(dp[i], dp[j] + 1);
        }
    }

    if (dp[s.size()] == INT_MAX)
        cout << "-1\n";
    else
        cout << dp[s.size()] << '\n';

    return 0;
}

File: saco/2019/dna/dna.py
import sys
sys.setrecursionlimit(10**9)
N, M = map(int, input().split())

dna = input()

dp = [float("inf") for i in range(N + 1)]
dp[N] = 0

class TrieNode:
    def __init__(self):
        self.children = {
            "A" : None,
            "C" : None,
            "G" : None,
            "T" : None
        }
        self.isEnd = False
    def add(self, s, i = 0):
        if (i == len(s)):
            self.isEnd = True
            return
        if (s[i] not in self.children) or (self.children[s[i]] is None):
            self.children[s[i]] = TrieNode()
        self.children[s[i]].add(s, i + 1)

root = TrieNode()
for i in range(M):
    x = int(input())
    root.add(input())

for k in range(N-1, -1, -1):
    dp[k] = float("inf")
    p = root
    for i in range(k + 1, N + 1):
        if (dna[i-1] not in p.children) or (p.children[dna[i-1]] is None):
            break
        p = p.children[dna[i-1]]
        if (p.isEnd):
            dp[k] = min(dp[k], dp[i] + 1)

if dp[0] == float("inf"):
    print(-1)
else: 
    print(dp[0])
